libevm/precompiles/p256verify/p256verify.go

```diff
@@ -0,0 +1,120 @@
+// them and/or modify them under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation, either version 3 of the License,
+// or (at your option) any later version.
+//
+// The libevm additions are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+// General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see
+// <http://www.gnu.org/licenses/>.
+
+// Package p256verify implements an EVM precompile to verify P256 ECDSA
+// signatures, as described in RIP-7212.
+package p256verify
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"io"
+	"math/big"
+
+	"github.com/ava-labs/libevm/params"
+)
+
+// Precompile implements ECDSA verification on the P256 curve, as defined by
+// [RIP-7212].
+//
+// [RIP-7212]: https://github.com/ethereum/RIPs/blob/1f55794f65caa4c4bb2b8d9bda7d713b8c734157/RIPS/rip-7212.md
+type Precompile struct{}
+
+// RequiredGas returns [params.P256VerifyGas].
+func (Precompile) RequiredGas([]byte) uint64 {
+	return params.P256VerifyGas
+}
+
+const inputLen = 160
+
+type input [inputLen]byte
+
+// Run parses and verifies the signature. On success it returns a 32-byte
+// big-endian representation of the number 1, otherwise it returns an empty
+// slice. The returned error is always nil.
+func (Precompile) Run(sig []byte) ([]byte, error) {
+	if len(sig) != inputLen || !(*input)(sig).verify() {
+		return nil, nil
+	}
+	return []byte{31: 1}, nil
+}
+
+func (in *input) verify() bool {
+	key, ok := in.pubkey()
+	if !ok {
+		return false
+	}
+	return ecdsa.Verify(key, in.word(0), in.bigWord(1), in.bigWord(2))
+}
+
+func (in *input) pubkey() (*ecdsa.PublicKey, bool) {
+	x := in.bigWord(3)
+	y := in.bigWord(4)
+	if x.Sign() == 0 && y.Sign() == 0 {
+		return nil, false
+	}
+
+	curve := elliptic.P256()
+	if !curve.IsOnCurve(x, y) {
+		return nil, false
+	}
+	return &ecdsa.PublicKey{
+		Curve: curve,
+		X:     x,
+		Y:     y,
+	}, true
+}
+
+func (in *input) word(index int) []byte {
+	s := index * 32
+	return in[s : s+32]
+}
+
+func (in *input) bigWord(index int) *big.Int {
+	return new(big.Int).SetBytes(in.word(index))
+}
+
+// Sign signs `hash` with the private key, using [rand.Reader] as the first
+// argument to [ecdsa.Sign]. It returns a signature payload constructed with
+// [Pack], which can therefore be passed directly to the precompile.
+func Sign(priv *ecdsa.PrivateKey, hash [32]byte) ([]byte, error) {
+	return signWithRandReader(rand.Reader, priv, hash)
+}
+
+// signWithRandReader is abstracted for testing purposes only and MUST NOT be
+// used directly. Always use [Sign].
+func signWithRandReader(rand io.Reader, priv *ecdsa.PrivateKey, hash [32]byte) ([]byte, error) {
```

Comment:
what's the benefit of having this method that accepts an `io.Reader`? Why not just hardcode it?